### DISCUSSION TITLE
Getting an error: "Failed to find Server Action "4a68...". This request might be from an older or newer deployment."

### DISCUSSION THREAD

**Discussion by Jared-Dahlke** [vercel/next.js#58431] [App Router]
I recently deployed an App dir 14.0.1 Nextjs application and have several users using it at my company. It uses server actions heavily. This AM i started getting some errors in Sentry:

```
Failed to find Server Action "4a682493280336616b4a9a906883df1c9dcdd5a5". This request might be from an older or newer deployment.
```

My assumption is that my users haven't refreshed their browser since I pushed out some changes.

Anyone ever seen anything like this? Any ideas or tips? Should I push out some change that makes users refresh when a new release is out?

**Comment by lotusjohn** [comment]
Getting the same in local dev on 14.0.0 up to 14.0.5 canary 5 for any server action which is triggered from client component. 
Server action in a server component works. 

**Working**: 
```
//ServerForm.tsx

export default function ServerForm() {
  const action = async (formData) => {
    'use server'
    console.log('server action', formData.get('name'))
  }
  return (
    <form action={action}>
      <input name={'name'} />
      <button type={'submit'}>Submit</button>
    </form>
  )
}
```

**Doesnt work**: 

```
//action.ts

'use server'
export async function formServerAction(params) {
  //code 
  return 'return value here'
}
```

```
//ClientCompForm.tsx

//component code
const onSubmit: SubmitHandler<FormInput> = async (data) => {
    //code
    await formServerAction(reqParams)
}
```

**Reply by Jared-Dahlke** [reply]
@lotusjohn, I started passing in server actions from server components and it seems to have fixed the problem.

so basically like this:

```
// server component

import { saveData } from 'actions/saveData'

export default async function Page() {
  

  return (
    <ClientComponent
      saveData={saveData}
    />
  );
}

```

Then i can use it in the ClientComponent like this:

```
'use client'

export const ClientComponent=({saveData})=> {

return (
  ...some form
<button onClick={saveData}></button>
)

}
```

or i can just keep the server action  in the server component like this:

```
// server component


export default async function Page() {
  
  const saveData=async(data)=> {
      'use server'
      //save data to db

   }

  return (
    <ClientComponent
      saveData={saveData}
    />
  );
}

```

Both seem to work, and prevent the error.


Can you confirm if you've tried similar? Glad to see i'm not the only one who experienced this.

**Reply by lotusjohn** [reply]
I just tried passing as prop as well and it worked!

**Comment by Jared-Dahlke** [comment]
@leerob this seems to be a problem, but we don't have the ability to make a repro repo because it requires multiple deployments to reproduce. can you please look into this or address?

**Comment by ken-spencer** [comment]
I am experiencing the same problem as well.  In  my application if I log out in one tab and then trigger a server action in the other tab I get this error every time.

**Reply by thomasrichmond** [reply]
Hey, I know this was a while ago, but wondering if you ever found a fix for this? We are using auth0 for user auth and the exact same happens when ever I trigger a server action on one tab after logging out out on another. Any help would be awesome as I am scratching my head over this!

**Reply by CodingDutchman** [reply]
Once your session is invalidated, wouldn't it be invalidated for all your tabs?

**Reply by thomasrichmond** [reply]
Yes it would be, but this was not the reason my FE was crashing. To make a long story short, triggering middleware from a server action (to perform auth checks using Next.redirect()) causes the server action to silently return undefined, causing a load of unexpected behaviour in your app. [Here](https://github.com/vercel/next.js/discussions/64993) is the thread explaining that hard auth checks should always be done at the request level, not the middleware level (complete opposite of the next docs!)

**Comment by mastoj** [comment]
I also got this now using next 14.0.4. For me it happens when the server actions is called from an parallel intercepted route. The exact same code works fine from the actual route.

**Reply by ken-spencer** [reply]
I was able to figure it out in my application. Logging out in one tab prevented the server actions in the other tab from working because the middleware was now redirecting them. Never would have guessed from the error :)

**Reply by ahmetkca** [reply]
I am facing the same problem. However, it is slightly different. I have a tanstack query in a client component where I use server action as `queryFn`. The problem arises when I invalidate this tanstack query from parallel intercepted route.

**Reply by Jared-Dahlke** [reply]
> I also got this now using next 14.0.4. For me it happens when the server actions is called from an parallel intercepted route. The exact same code works fine from the actual route.

Can you please expand on this? What do you mean by "parallel intercepted route" and "actual route"? Explain like I am 80 IQ please. Can you give example? Why is this not being fixed or better addressed?

**Reply by ahmetkca** [reply]
> > I also got this now using next 14.0.4. For me it happens when the server actions is called from an parallel intercepted route. The exact same code works fine from the actual route.
> 
> Can you please expand on this? What do you mean by "parallel intercepted route" and "actual route"? Explain like I am 80 IQ please. Can you give example? Why is this not being fixed or better addressed?

What I mean by parallel intercepted route is this:
<img width="328" alt="Screenshot 2024-01-14 at 2 31 39 PM" src="https://github.com/vercel/next.js/assets/74574469/af6e32a8-6d0f-44d3-9fa4-0e12a644edbe">


In `/boards/[boardId]` page.tsx I display tasks for the given board. If you click on a task, it opens the comments for the clicked task in the modal, `/boards/[boardId]/tasks/[taskId]/comments`. 

The route `/boards/[boardId]/tasks/[taskId]/comments` is rendered on top of `/boards/[boardId]` in a modal. So the route `/boards/[boardId]/tasks/[taskId]/comments` is being intercepted.

User adds a a new comment. I want to invalidate the query so that the number of comments in the `/boards/[boardId]` updates. I invalidate the tanstack query but because server actions cannot be found it fails.
<img width="238" alt="Screenshot 2024-01-14 at 2 37 51 PM" src="https://github.com/vercel/next.js/assets/74574469/88c61269-3003-49cd-8dbe-b4b065dd11b7">

```
Error: Failed to find Server Action "de6c6489e15944f0b88e9d483d2eef1d967c7a4c". This request might be from an older or newer deployment. Original error: Invariant: Couldn't find action module ID from module map.
```

**Comment by mkbctrl** [comment]
In my case it happens when trying to redirect on successfully submitted form. 
Fun fact: it worked for like an hour or two, and then it started to throw error 🙃 - maybe a local dev issue?
```
useEffect(() => {
    if (state?.submitted) {
      actionRedirect(Path.UserHome)
    }
  }, [state])
```

where:
```
'use server'

import { Path } from 'src/constants/path'
import { redirect } from 'src/services/i18n/i18n-navigation'

/**
 * actionRedirect - if you need to trigger redirect from the client component
 * https://nextjs.org/docs/app/api-reference/functions/redirect
 */
export async function actionRedirect(path: Path) {
  redirect(path)
}

```

error:
```
rror: Failed to find Server Action "017bd78c23c5fecaef8cb9b835f35b17bdd6640d". This request might be from an older or newer deployment. Original error: Invariant: Couldn't find action module ID from module map.
```

By the way, anyone has an approach to onFormSubmission side effects, and is willing to share? I am working with native useFormState, gave it a ride, and it's missing a bit in flexibility or I can't wrap my head around on Next.js 14 specifics yet

EDIT:
In my case non of the workarounds work.

EDIT2:
Using router.push started to work -> I ended up reading this https://nextjs.org/docs/app/api-reference/functions/redirect due to the warning the contents I don't remember stating I should use server action... What a maze 😂

**Comment by paulinhapenedo** [comment]
I'm getting the same error. The full error message:

```bash
Error: Failed to find Server Action "ff338b32de2c62c9fce3998a34545a3e0672f704". This request might be from an older or newer deployment. Original error: Invariant: Couldn't find action module ID from module map.
```

The stack I'm using:
**next**: 14.1.0
**next-auth**: 4.24.5
**react-query**: 5.18.0

It happens whenever the user's JWT token expires and Tanstack Query tries to refetch the data on the background. All of my query/mutation functions are Server Actions. In them, I have this code running before actually fetching anything:

```ts
// serverAuth lives in another file
export const serverAuth = async () => await getServerSession(authOptions);


// actual Server Action
'use server';

const getUnitPractitioners = () => {
 const session = await serverAuth();

  if (!session) {
    LOG.error('🚨 No session found when fetching practitioners for this unit');
    return Promise.reject(
      'No session found when fetching practitioners for this unit'
    );
  }

// does whatever action is necessary

return data;
}
```

**Comment by steventkrawczyk** [comment]
I'm experiencing the same issue

**Comment by ghunkins** [comment]
Same issue on both stable and canary releases for `next`.

Re-echoing that it appears to occur when using the [Client Components](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations#client-components) data fetching paradigm.

**Reply by bernaferrari** [reply]
You are right but same would happen with api routes, like, if you changed a route it would still reach the old one. I guess it is less bad, but would still happen.

**Comment by Lebraz98** [comment]
I have same problem.
I have fixed it by remove request.headers from  fetch header and create new header value
not pass the next request header inside fetch create new header request value like this 

old code: [image](https://github.com/vercel/next.js/assets/59622265/a4975157-0f32-491d-9232-e4617ebfd0dd)

new code: 
![image](https://github.com/vercel/next.js/assets/59622265/bca9e167-3859-45bc-afda-2716bbd12283)

**Comment by Prains** [comment]
Same issue here. Happens when i reopen intercepted modal window that uses server actions in client components using browser navigation.

**Comment by teddybee** [comment]
Same issue here. 
I would like to do a simple logout (jwt cookie clear) and redirect to home in server action.

**Reply by newtonMM** [reply]
I'm experiencing the same issue, were you able to fix ir

**Comment by GVALFER** [comment]
Same here

**"next": "^14.1.0",**

```js
Error: Failed to find Server Action "572c475e1c59eeb141374e2737137c179e2063dc". 
This request might be from an older or newer deployment. Original error: Invariant: Couldn't find action module ID from module map.
    at t3 (/.../node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:39:1693)
    at /.../node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:38:7079
    at AsyncLocalStorage.run (node:async_hooks:338:14)

**Comment by abdtriedcoding** [comment]
Any updates facing same issue ?

**Comment by RestartDK** [comment]
I get a similar error when I use react-email...

**Comment by sim391calado** [comment]
I have this exact same issue.

**Comment by jaz-creative** [comment]
I have the same issue, using the workaround of passing the action from the server component to client as a prop.

**Comment by TheLexoPlexx** [comment]
For me, this problem appeared, when I put a Server Component within a Client Component and had to explicitly tell the file "use server" so it executes on the server.

**Comment by eduardodallmann** [comment]
I have a slightly different problem.

I will describe how my application works.
It has a listing of events. When I click edit, I navigate to events/[slug]. The form will use a server action with debouncing to save. To control this debouncing I use context api, also to show on the screen that it is saving. After making an edit to the form and waiting 2 seconds, the data is saved and the /events path is revalidated. Everything works perfectly.

Now I will describe the problem.
When I edit the form and click on the backdrop quickly before 2 seconds, the drawer is closed and the application navigates to /events again. When the 2 seconds are complete, the terminal displays the error that the server action was not found.

Originally I didn't use server actions, I was using a client function with axios accessing an api path in the nextjs application itself. He worked normally. I was using this application to study server actions and I came across this problem.

Recording:
![server action problem](https://github.com/vercel/next.js/assets/7562936/48978027-4068-4147-829e-a3db86667fba)

[Codesandbox](https://codesandbox.io/p/github/eduardodallmann/app-next-server-action-problem)

**Reply by sim391calado** [reply]
I believe this is the same issue that has been reported here:
#62102

**Comment by nemanjaonex2** [comment]
It seems I solved this by moving server actions and validation schema definitions outside of `/app` folder.

```ts
Error: Failed to find Server Action "...". This request might be from an older or newer deployment. 
Original error: Cannot read properties of undefined (reading 'workers')

// solution:
server action and shared validation must be defined OUTSIDE of /app folder
```

**Reply by jvmdo** [reply]
It worked for me.
I got the error trying to create the container using the [official method](https://github.com/vercel/next.js/tree/canary/examples/with-docker#in-existing-projects) but no errors when directly running `npm run build`.

**Comment by CodingDutchman** [comment]
Thank you all for your suggestions. After hours of trying everything I could find online, I finally remembered that I had added the following lines to my next-auth config.
```
  session: {
    maxAge: 60, // 1 minute, for testing
  },
```
Next-auth was thus automatically invalidating my session after 60 seconds and the client-side could not reach the server actions. Once I deleted these lines from the next-auth config, the client side no longer had issues finding the Server Actions.

So if your server actions are working for awhile, but then you get a "failed to find Server Action", double check to make sure your session is still active.

Hopefully this saves someone else some time and makes up for all the time I spent. :-P

**Reply by ezraanglo** [reply]
i don't understand how this worked but the error is no more so thanks!
Tho i don't have that line and i had to add 
```
maxAge: 30 * 24 * 60 * 60, // 30 days
```
which is supposed to be the default. im using the next-auth@beta on this so it might be a bug on their side.

**Comment by AndonMitev** [comment]
Hello i'm experiencing the same problem on production 

<img width="1110" alt="image" src="https://github.com/vercel/next.js/assets/32274987/858f911b-a204-4ca0-b620-fb0c4704df51">


Only on prod tho, i've deployed without using cache etc but issue persists, any ideas what can be done?

**Comment by Artem-Semenov** [comment]
The same problem on trying to revalidate path. Passing server action as prop didnt work for me. Any solutions? I need to render different ui for different user roles - my sidebar is server component and if I log out from one user role and log in with another - my sidebar continues to be stale and refreshes only after reloading the page. With revalidating path It refreshes and then after logging under another user role I see for d 0.5s stale sidebar and then it refreshed to the appropriate one which is also not desired behavior + in that case I am getting this error - "Failed to find Server Action "4a682493280336616b4a9a906883df1c9dcdd5a5". This request might be from an olde bla bla bla"

**Comment by SSTPIERRE2** [comment]
My NextJS app has this error whenever I trigger a server action on one page, then navigate to a different page where that action doesn't exist.

**Comment by kassiogluten** [comment]
Same problem here

**Comment by AhmedAlbarghouti** [comment]
Running into the same issue too.

I have a server action called from a client component (server action passed in from server component) which is checked by the middleware if rate limit is reached then prompting a redirect if limit is reached.

If I swap out the redirect and return a json response and deal with the redirection on the client side, then the issue gets resolved.
I guess it can't find the server action after redirection.

I guess I can use this workaround for now, but it would be nice if we can get something from the Vercel team for this.

**Comment by tuon1602** [comment]
I have an issue "failed to find server action".before it logs the error, i can call server actions 2 times and it's dies when trigger the 3rd time, idk why

**Reply by tuon1602** [reply]
thinks i have tried, update longer maxAge for next-auth
move actions folder outside of the app folder

**Reply by tuon1602** [reply]
I fixed it by upgrading from my current version 14.4.1 to newest nextjs version ( currently 14.2.0 ) , guys should try it

**Reply by gitNejc** [reply]
Same here

**Comment by Jared-Dahlke** [comment]
Yeah honestly I would recommend just use TRPC for all of your server calls. It will create api routes for all of them. APi routes are just more stable and predictable

**Comment by T3chnoMaciej** [comment]
Hello, for anyone that may encounter similar problem - in my case it was dynamic og-image generation that caused "Failed to find server action" error. NextJS changed package name from "next/server" to "next/og". That small change fixed error for my use case with dynamic opengraph.

From:
``
import { ImageResponse } from "next/server";
``
To:
``
import { ImageResponse } from "next/og";
``

I hope it saves someone's time. Cheers

**Comment by Oteiza-a** [comment]
This error happens to me when I call a server action in my useEffect in a client component, **and do a client-side navigation to another page before the server action finishes**. This results in the client receiving `undefined` as response and the server side generating the `Error: Failed to find Server Action` error. 

For now I will just validate that the server action responds a valid value, but I still find no solution for the server side.

**Comment by OU9999** [comment]
in my case, upgrading the version of `next` worked well in my case.

error message
```
Error: Failed to find Server Action "~~~". This request might be from an older or newer deployment
```

before
```
"next": "14.1.3",
```

after
```
 "next": "^14.2.5",
```

**Comment by noorulh06** [comment]
Any update on this?
I am on latest version of next js "^14.2.5" and getting same issue.
Error: Failed to find Server Action "74c865cbe0806fdeefdb08160fa74838f6ea96ff".......
No matter what I try, server action always generates new id on each build. Though I don't know why it's not reproducible locally.

**Comment by adriangalilea** [comment]
Following: https://github.com/vercel/next.js/pull/69547

> I'm going ahead and shipping this one because it's quite literally broken on main.
@leerob 

I tried the versions on the example here

```
    "react": "rc",
    "react-dom": "rc",
    "next": "canary"
```

And I ran into similar issues, `Error: Failed to find Server Action ...`...

I also found literally impossible to use server actions inside client components to fetch data.

None of the methods I used before worked for me, and I get no proper error/warning so I have no clue.

Tried:
- SWR
- TanstackQuery
- useEffect

All methods that I used previously and worked to some degree failed, so I have no idea how to use server actions from client components as of now.

**Reply by probablykabari** [reply]
This started happening to me in next 14.2.8. Downgrading to 14.2.7 resolved it for the time being.

**Reply by Suprndm** [reply]
Same

**Reply by joel-wlf** [reply]
Same here

**Comment by controversial** [comment]
I’m also seeing this error appear on 14.2.8, even when on the latest deployment

**Reply by controversial** [reply]
I found a reproduction and filed https://github.com/vercel/next.js/issues/69756

**Reply by controversial** [reply]
cc @probablykabari

**Comment by controversial** [comment]
#69788 now merged to fix the issue that was introduced in 14.2.8

**Comment by chriskuech** [comment]
I'm observing this issue with 14.2.9

**Reply by controversial** [reply]
can you create a reproduction and send it to @huozhi in #69756?

**Reply by chriskuech** [reply]
It turns out the issue was that I had a normal helper function that contained `use client` at the top of the file. It was added by the VS Code "Refactor" feature so I did not realize it had been added.

I would very much appreciate stronger static validation of Next apps as cryptic build and runtime errors are all too common and extra hard to debug with inaccurate error messages.

**Comment by fi5u** [comment]
At least for me, was still broken with 14.2.9 but now issue is fixed with 14.2.10.

**Comment by huozhi** [comment]
We released v14.2.10 with the fix, could you upgrade to the latest version and verify if that fixed for you? 🙏 Thanks

**Comment by arman94** [comment]
I encountered the same issue. I was importing my server action using an absolute path, but switching to a relative import resolved the problem

**Comment by huozhi** [comment]
If you still encounter it with latest Next.js version, please file an issue with reproduction, we can help take a look on it! 🙏

**Reply by marsonmao-trait** [reply]
Hi @huozhi, just wondering if we're expected to use the practice in https://github.com/vercel/next.js/discussions/58431#discussioncomment-7620263 (i.e. using props to pass server actions around), or are we good with just importing them from a file in a Client component? 🙏

**Reply by huozhi** [reply]
@marsonmao-trait It should work if you import server action into client components, if you still have issues with it probably a bug. Would you mind filing an issue with your repro and tag me there? Thanks

**Comment by marsonmao-trait** [comment]
I installed 14.2.10 and I'm still getting `undefined` with a un-refreshed browser tab calling a server action.

Scenario: a user opens a tab, and then the server has a new deployment while the tab is not refreshed i.e. it becomes outdated / stale; in this case if the user Submit (with server action) using the stale client, the server action returns `undefined`

----

I tried this pattern: import server actions using relative path in a server component, and pass it as a prop to a client component, but it does not work

Referenced from https://github.com/vercel/next.js/discussions/58431#discussioncomment-7620263 and https://github.com/vercel/next.js/discussions/58431#discussioncomment-10635926

----

The specific behavior on my un-refreshed tab is: when clicking Submit, it fired a POST server action with header `Accept = text/x-component`, but and then it received a 200 response, with header `Content-type = text/html; charset=utf-8`, also the server action returns `undefined`.

So from the Dev console we can see an uncaught error saying "reading xxx from undefined", also the user would see no changes from the page (because the response seems to be the exact HTML structure of the current page).

In this case if the user chooses to refresh the tab, then it can work, but not every time the users are aware of this, obviously.

So I'm wondering how can I maybe detect this kind of response (server action returning `undefined`) gracefully, i.e. not to try-catch on "every" server action call, so that I can show an error message to the users and propose a manual refresh?

**Comment by abhi12299** [comment]
from the nextjs docs: https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations#security:~:text=However%2C%20for%20this,a%20specific%20build.
it seems that a server action can only be invoked for a particular build.

If this is the constraint, we can work around this by having an api route serve the build id to the frontend, and the frontend can poll for the build id. If it sees any changes, we can ask the user to refresh the page

**Reply by benjick** [reply]
@abhi12299 got an example? 😇

**Reply by abhi12299** [reply]
@benjick haven't tried it myself, but you can serve the value of `process.env.GIT_HASH` on an API route to check for staleness. Ref: https://nextjs.org/docs/app/api-reference/config/next-config-js/generateBuildId

If that is not available after the build time, you can modify the webpack/turbopack config in next.config.js/ts file

**Comment by benjick** [comment]
This feature of generating new "unguessable, non-deterministic IDs" for server actions on every build is ruining the ability to continuously deploy my application. I don't want to force my users to refresh their browsers every time I do a deploy. What's the solution for self-hosters? Can this feature be turned off?

**Reply by controversial** [reply]
I believe you can [set `NEXT_SERVER_ACTIONS_ENCRYPTION_KEY`](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations#overwriting-encryption-keys-advanced) to a fixed value

**Reply by benjick** [reply]
@controversial thanks, I'll try it! I was just thinking how it would be nice to set a key yourself and rotate it on demand and came back to this thread to suggest it. Cheers!

**Reply by benjick** [reply]
I've been reading through this issue: https://github.com/vercel/next.js/issues/75541

But it's not really clear if/when the id could change even after setting `NEXT_SERVER_ACTIONS_ENCRYPTION_KEY`

**Reply by benjick** [reply]
It doesn't work, at least not for me. Did a deploy and an action which is in a file that hasn't been changed for a week failed.

**Comment by wieringen** [comment]
We are self-hosting Next.js on Cloud Run. We have set the NEXT_SERVER_ACTIONS_ENCRYPTION_KEY, but that made no difference. This error is still popping up all the time. Is there anything that can be done?

**Reply by benjick** [reply]
@leerob will this be solved? Or is it solvable in another way?

**Reply by Davydhh** [reply]
Any news?

**Reply by benjick** [reply]
I got it working, set it both during build and runtime

**Reply by notflip** [reply]
@wieringen What value did you put in NEXT_SERVER_ACTIONS_ENCRYPTION_KEY? Is it just a random string?

**Reply by wieringen** [reply]
```
resource "random_bytes" "next_encryption" {
  length = 32
}
```
I used the terraform random_bytes resource

**Reply by benjick** [reply]
How do you use it?

**Reply by IkerPagoaga** [reply]
> I got it working, set it both during build and runtime

Do you set it in a docker building script and then store it on a secret?

**Reply by benjick** [reply]
@IkerPagoaga what do you mean store it on a secret?
1. Pass it to `next build` as an env var via build secret in my Dockerfile
2. Set it as an env var during runtime

**Reply by IkerPagoaga** [reply]
> @IkerPagoaga what do you mean store it on a secret?
> 
> 1. Pass it to `next build` as an env var via build secret in my Dockerfile
> 2. Set it as an env var during runtime

I found out it can be troubling if you hardcode the value, so just as yourself, I add it to my dockerfile without being hardcoded. If you use google cloud secret manager you can forget about having to update the key value itself since you are using the newest latest version of the secret that gets created with a different script. (Edit: only 6 secrets can be kept active if you dont want to start paying for them on gc secret manager, if you dont need to store them, store and destroy the previous version. Update this value each 3 months, you shouldnt have to worry about it.)

**Comment by hessaam** [comment]
I toggled client component to the server component (APP CRASHED) and to the client component again (!!) and everything had been OK.

**Comment by AdarshaNayak** [comment]
We have set the NEXT_SERVER_ACTIONS_ENCRYPTION_KEY but still we run into `Failed to find server action` error. Although the frequency has reduced we would not like to have it in the first place. 
Is there a way to disable non-deterministic nature of these IDs?
I'm currently on v15.3.4

**Comment by markedwards** [comment]
Also running into this on 14.2.33, after switching to a standalone build.  What is the deal with this?  This is severely breaking our UX.

**Reply by markedwards** [reply]
Update on this -- it turned out in our case to be non-deterministic absolute paths in CI causing server action ids to change on every deploy.  Fixing that solved the issue.

**Reply by kyrylolvov** [reply]
Could you elaborate on this? How were you able to fix it?

**Reply by markedwards** [reply]
So in our case, we switched to doing the `next build` directly inside a GitHub runner, instead of inside of a docker image build.  GitHub actions have a different absolute path for `$GITHUB_WORKSPACE` on every run.  So that was causing the `next build` to generate a new id for every action on every run.

The fix was to do:
```
mkdir /company
mount --bind "$GITHUB_WORKSPACE" /company
```

And then perform our pipelines steps from the `/company` directory instead of `$GITHUB_WORKSPACE`.  However, this is only possible because we use AWS Codebuild as our custom runner.  I don't think you are allowed to do the above in a standard GitHub runner.

TLDR -- you have to have a deterministic absolute path on every build in your CI, somehow.

**Comment by kyrylolvov** [comment]
Currently facing this issue on 15.1.4, hosting in Docker on ECS

**Comment by leandronn** [comment]
For me it was a browser issue. Hard reloading the browser, worked just fine.

**Reply by markedwards** [reply]
I think you might misunderstand.  That's exactly what's being complained about here.  The app should continue working without the user refreshing.

**Reply by developerEhsan** [reply]
Guys did you find any solution?

**Reply by leandronn** [reply]
> I think you might misunderstand. That's exactly what's being complained about here. The app should continue working without the user refreshing.

but if you rebuild your .next compilation, browser will have cached "old" action IDs. the app should not continue working with a diferent build.

**Comment by kyrylolvov** [comment]
I feel like for me the issue is that there's other stuff running on the ECS, and whenever those get deployed, they restart Next.js application, which seems to break Server Actions. Might be wrong though. Basically, Next.js needs like ideal environment with nothing interrupting it.

**Comment by JoshuaSmeda** [comment]
Getting this issue at all the time on Next `16.0.10` on Dokploy.

I've even set `NEXT_SERVER_ACTIONS_ENCRYPTION_KEY` but makes no difference.

**Comment by youssefKetata** [comment]
Did someone figure out a solution for this , without using client functions?

**Reply by kyrylolvov** [reply]
Moved to API routes

**Reply by youssefKetata** [reply]
I tried it, did not work for me, same error appers

**Reply by Pratham5236** [reply]
Anyone could figure this out? Its bugging a lot of our users

**Comment by anjan-kp** [comment]
Does this issue occur if we don't use any server action? 
I am using NextJS 15.4.10
How about adding generateBuildId to next.config.ts?

**Comment by Pratham5236** [comment]
To anyone facing this issue, while self hosting Next.js, please try adding NEXT_DEPLOYMENT_ID or deploymentId to next.config.ts/js. According to official documentation [here](https://nextjs.org/docs/app/api-reference/config/next-config-js/deploymentId), this helps in version skewing protection. I have added this and so far, I'm not seeing issues as i was earlier
